pkg/sdk/go/sdk.go

```diff
@@ -353,3 +358,37 @@ func (sdk mfSDK) sendThingRequest(req *http.Request, key, contentType string) (*
 
 	return sdk.client.Do(req)
 }
+
+func (sdk mfSDK) withQueryParams(baseURL, endpoint string, pm PageMetadata) (string, error) {
+	q, err := pm.Query()
+	if err != nil {
+		return "", err
+	}
+	return fmt.Sprintf("%s/%s?%s", baseURL, endpoint, q), nil
+}
+
+func (pm PageMetadata) Query() (string, error) {
```

Comment:
Please double-check this implementation. There is a `Values` type from `net/url` that you can use for this purpose, rather than manually constructing a query. It is supposed to provide you a better way of handling query and escaping it correctly. Consider something like:
```go
func (pm PageMetadata) Query() (string, error) {
    var resp map[string]interface{}
    jsonMarshal, err := json.Marshal(pm)
    if err != nil {
    	return "", err
    }
    err = json.Unmarshal(jsonMarshal, &resp)
    if err != nil {
    	return "", err
    }
    q := url.Values{}
    for key, val := range resp {
    	var v string
    	ok := false
    	if key != "metadata" {
    		v, ok = val.(string)
    		if !ok {
    			return "", ErrMetadata
    		}
    	}
    	switch key {
    	case "name", "email":
    		q.Add(key, v)
    	case "total", "offset", "limit":
    		if _, err := strconv.ParseUint(v, 10, 64); err != nil {
    			return "", err
    		}
    		q.Add(key, v)
    	case "metadata":
    		md, err := json.Marshal(pm.Metadata)
    		if err != nil {
    			return "", err
    		}
    		q.Add(key, string(md))
    	}
    }
    return q.Encode(), nil
}
```

Reply:
Since in pagemetadata only email and name are string, this raises an error
```
  v, ok = val.(string)
  if !ok {
   return "", error
  }
```
I have changed it to just use `fmt.Sprint(val)` for all values apart from pageMetadata

Reply:
`Since in pagemetadata only email and name are string, this raises an error` Makes sense. Since we do not get many benefits from using this marshaling to JSON and back just to remove unused fields while also losing type information, I would propose sticking to the original PageMetadata and simple `if`:
```go
func (pm PageMetadata) query() (string, error) {
	q := url.Values{}
	if pm.Level != 0 {
		q.Add("level", strconv.FormatUint(pm.Level, 10))
	}
	if pm.Total != 0 {
		q.Add("total", strconv.FormatUint(pm.Total, 10))
	}
	if pm.Offset != 0 {
		q.Add("offset", strconv.FormatUint(pm.Offset, 10))
	}
	if pm.Limit != 0 {
		q.Add("limit", strconv.FormatUint(pm.Limit, 10))
	}
	if pm.Level != 0 {
		q.Add("level", strconv.FormatUint(pm.Level, 10))
	}
	if pm.Email != "" {
		q.Add("email", pm.Email)
	}
	if pm.Name != "" {
		q.Add("name", pm.Name)
	}
	if pm.Type != "" {
		q.Add("type", pm.Type)
	}
	if pm.Metadata != nil {
		md, err := json.Marshal(pm.Metadata)
		if err != nil {
			return "", err
		}
		q.Add("metadata", string(md))
	}
	return q.Encode(), nil
}
```

Reply:
We don't retain a value of zero for uint fields. For instance, if we pass in a limit of 0 it should return an `ErrFailedFetch`. I have changed it to allow that.

Reply:
@0x6f736f646f Great, even simpler. Only keep this method unexported, we use it only locally.